aioshelly/wsrpc.py

```diff
@@ -0,0 +1,194 @@
+"""WsRpc for Shelly."""
+from __future__ import annotations
+
+import asyncio
+import logging
+from dataclasses import dataclass
+from datetime import datetime
+from typing import Any, Dict
+
+from aiohttp import ClientWebSocketResponse, client_exceptions
+
+from .const import WS_HEARTBEAT_SEC
+from .exceptions import (
+    CannotConnect,
+    InvalidMessage,
+    JSONRPCError,
+    RPCError,
+    RPCTimeout,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass
+class RPCCall:
+    """RPCCall class."""
+
+    call_id: int
+    method: str
+    params: Dict[str, Any] | None
+    src: str | None = None
+    dst: str | None = None
+    sent_at: datetime | None = None
+    resolve: asyncio.Future | None = None
+
+    @property
+    def request_frame(self):
+        """Request frame."""
+        msg = {
+            "id": self.call_id,
+            "method": self.method,
+            "src": self.src,
+        }
+        for obj in ("params", "dst"):
+            if getattr(self, obj) is not None:
+                msg[obj] = getattr(self, obj)
+        return msg
+
+
+class WsRPC:
+    """WsRPC class."""
+
+    def __init__(self, ip_address: str, on_notification):
+        """Initialize WsRPC class."""
+        self._ip_address = ip_address
+        self._on_notification = on_notification
+        self._rx_task = None
+        self._websocket: ClientWebSocketResponse | None = None
+        self._calls: Dict[str, str] = {}
+        self._call_id = 1
+        self._src = f"aios-{id(self)}"
+        self._dst = None
+
+    @property
+    def _next_id(self):
+        self._call_id += 1
+        return self._call_id
+
+    async def connect(self, aiohttp_session):
+        """Connect to device."""
+        if self.connected:
+            _LOGGER.debug("%s already connected", self._ip_address)
+            return
+
+        _LOGGER.debug("Trying to connect to device at %s", self._ip_address)
+        try:
+            self._websocket = await aiohttp_session.ws_connect(
+                f"http://{self._ip_address}/rpc", heartbeat=WS_HEARTBEAT_SEC
```

Comment:
In aiohttp the heartbeat is broken. It will send a ping but if it times out it won't cancel any open receive command. If the connection is closed it will print a stack trace that's not tracked. 

Instead I would follow this approach:
https://github.com/NabuCasa/hass-nabucasa/blob/master/hass_nabucasa/iot_base.py#L180-L185